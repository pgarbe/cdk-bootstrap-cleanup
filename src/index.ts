import * as cdk from 'aws-cdk-lib';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct, IConstruct } from 'constructs';

export interface BootstrapCleanupProps {

  readonly bucket: string;
  readonly key: string;
}

/**
 * Construct to clean up a given asset
 */
export class BootstrapCleanup extends Construct {
  /**
   *
   */
  constructor(scope: Construct, id: string, props: BootstrapCleanupProps) {
    super(scope, id);

    new cdk.custom_resources.AwsCustomResource(this, 'Cleanup', {
      installLatestAwsSdk: false,
      onUpdate: {
        service: 'STS',
        action: 'getCallerIdentity',
        physicalResourceId: cr.PhysicalResourceId.of(`${props.bucket}/${props.key}`), // Update physical id to always fetch the latest version
      },
      onDelete: {
        service: 'S3',
        action: 'deleteObject',
        parameters: {
          Bucket: props.bucket,
          Key: props.key,
        },
      },
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
        resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}

export class BootstrapCleanupAspect implements cdk.IAspect {

  visit(node: IConstruct): void {
    if (node instanceof cdk.aws_s3_assets.Asset) {
      new BootstrapCleanup(node, `${node.node.id}-cleanup`, { bucket: node.s3BucketName, key: node.s3ObjectKey });
    }
  }
}
