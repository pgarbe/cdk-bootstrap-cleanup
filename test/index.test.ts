import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { BootstrapCleanupAspect } from '../src';

test('Defaults are correctly set', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  // WHEN
  new cdk.aws_s3_assets.Asset(stack, 'S3Asset', {
    path: 'src',
  });

  cdk.Aspects.of(app).add(new BootstrapCleanupAspect());

  // THEN
  // Repository is created
  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});