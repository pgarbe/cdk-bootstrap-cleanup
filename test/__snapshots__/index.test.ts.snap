// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defaults are correctly set 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "S3AssetS3AssetcleanupCleanup50EC0451": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "S3AssetS3AssetcleanupCleanupCustomResourcePolicyA16CEEE3",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"STS\\",\\"action\\":\\"getCallerIdentity\\",\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/67112e5ff5d3625c22aa8596601f13df219842137f67b26be9ff15dd67ee2016.zip\\"}}",
            ],
          ],
        },
        "Delete": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"S3\\",\\"action\\":\\"deleteObject\\",\\"parameters\\":{\\"Bucket\\":\\"",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "\\",\\"Key\\":\\"67112e5ff5d3625c22aa8596601f13df219842137f67b26be9ff15dd67ee2016.zip\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"STS\\",\\"action\\":\\"getCallerIdentity\\",\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/67112e5ff5d3625c22aa8596601f13df219842137f67b26be9ff15dd67ee2016.zip\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "S3AssetS3AssetcleanupCleanupCustomResourcePolicyA16CEEE3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:GetCallerIdentity",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "s3:DeleteObject",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3AssetS3AssetcleanupCleanupCustomResourcePolicyA16CEEE3",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
